llama_stack/providers/inline/post_training/huggingface/recipes/finetune_single_device.py

```diff
@@ -0,0 +1,423 @@
+        lr = 2e-5
+        if config.optimizer_config:
+            lr = config.optimizer_config.lr
+        # Split dataset into train and validation
+        train_val_split = ds.train_test_split(test_size=0.1, seed=42)
+        train_dataset = train_val_split["train"]
+        eval_dataset = train_val_split["test"]
+
+        # Calculate steps per epoch and appropriate intervals
+        steps_per_epoch = len(train_dataset) // config.data_config.batch_size
+        eval_steps = max(1, steps_per_epoch // 10)  # Evaluate 10 times per epoch
+        save_steps = max(1, steps_per_epoch // 5)  # Save 5 times per epoch
+        logging_steps = max(1, steps_per_epoch // 50)  # Log 50 times per epoch
+
+        logger.info(f"Dataset size: {len(train_dataset)} examples")
+        logger.info(f"Batch size: {config.data_config.batch_size}")
+        logger.info(f"Steps per epoch: {steps_per_epoch}")
+        logger.info(f"Will evaluate every {eval_steps} steps")
+        logger.info(f"Will save every {save_steps} steps")
+        logger.info(f"Will log every {logging_steps} steps")
+
+        training_arguments = SFTConfig(
+            max_steps=config.max_steps_per_epoch,
+            output_dir=str(output_dir_path),
+            num_train_epochs=config.n_epochs,
+            per_device_train_batch_size=config.data_config.batch_size,
+            fp16=device.type == "cuda",
+            bf16=device.type != "cuda",
+            use_cpu=device.type == "cpu",
+            save_strategy="steps",
+            save_steps=save_steps,
+            report_to="none",
+            max_seq_length=provider_config.max_seq_length,
+            gradient_accumulation_steps=config.gradient_accumulation_steps,
+            gradient_checkpointing=provider_config.gradient_checkpointing,
+            learning_rate=lr,
+            warmup_ratio=provider_config.warmup_ratio,
+            weight_decay=provider_config.weight_decay,
+            logging_steps=logging_steps,
+            # Enable validation
+            eval_strategy="steps",
+            eval_steps=eval_steps,
+            save_total_limit=provider_config.save_total_limit,
+            remove_unused_columns=False,
+            dataloader_pin_memory=provider_config.dataloader_pin_memory,
+            dataloader_num_workers=provider_config.dataloader_num_workers,
+            dataset_text_field="text",
+            packing=False,
+            # Add evaluation metrics
+            load_best_model_at_end=True,
+            metric_for_best_model="eval_loss",
+            greater_is_better=False,
+        )
+
+        # Initialize trainer with both train and eval datasets
+        trainer = SFTTrainer(
+            model=model_obj,
+            train_dataset=train_dataset,
+            eval_dataset=eval_dataset,
+            peft_config=peft_config,
+            args=training_arguments,
+        )
+
+        # Train
+        logger.info("Starting training")
+        try:
+            trainer.train()
```

Comment:
this line blocks shutdown of the server process because it never yields back. A correct implementation would yield control periodically to allow loop.stop to execute.

AFAIU there's no easy way to make transformers library to do so. In this case, an alternative way to yield will have to be devised, e.g. run the trainer in a subprocess; then in your "main" routine wait for process execution but otherwise yield every so often.

In the future, scheduler will pass a callback function into your handler that you would call to check if shutdown is happening and perhaps subprocess.terminate(). This will have to wait.

Reply:
resolved in this PR, shutdown is now handled gracefully